bottlerocket/agents/src/bin/ec2-karpenter-resource-agent/ec2_karpenter_provider.rs

```diff
@@ -0,0 +1,982 @@
+    }
+
+    info!("Creating KarpenterInstanceProfile");
+    iam_client
+        .create_instance_profile()
+        .instance_profile_name("KarpenterInstanceProfile")
+        .send()
+        .await
+        .context(Resources::Clear, "Unable to create instance profile")?;
+
+    iam_client
+        .add_role_to_instance_profile()
+        .instance_profile_name("KarpenterInstanceProfile")
+        .role_name("KarpenterInstanceNodeRole")
+        .send()
+        .await
+        .context(Resources::Clear, "Unable to add role to InstanceProfile")?;
+
+    Ok(())
+}
+
+async fn create_controller_policy(
+    iam_client: &aws_sdk_iam::Client,
+    account_id: &str,
+) -> ProviderResult<()> {
+    if iam_client
+        .get_policy()
+        .policy_arn(format!(
+            "arn:aws:iam::{}:policy/KarpenterControllerPolicy",
+            account_id
+        ))
+        .send()
+        .await
+        .is_ok()
+    {
+        info!("KarpenterControllerPolicy already exists");
+        return Ok(());
+    }
+
+    info!("Creating controller policy");
+    iam_client
+        .create_policy()
+        .policy_name("KarpenterControllerPolicy")
+        .policy_document(
+            r#"{
+        "Statement": [
+            {
+                "Action": [
+                    "ssm:GetParameter",
+                    "iam:PassRole",
+                    "ec2:DescribeImages",
+                    "ec2:RunInstances",
+                    "ec2:DescribeSubnets",
+                    "ec2:DescribeSecurityGroups",
+                    "ec2:DescribeLaunchTemplates",
+                    "ec2:DescribeInstances",
+                    "ec2:DescribeInstanceTypes",
+                    "ec2:DescribeInstanceTypeOfferings",
+                    "ec2:DescribeAvailabilityZones",
+                    "ec2:DeleteLaunchTemplate",
+                    "ec2:CreateTags",
+                    "ec2:CreateLaunchTemplate",
+                    "ec2:CreateFleet",
+                    "ec2:DescribeSpotPriceHistory",
+                    "pricing:GetProducts"
+                ],
+                "Effect": "Allow",
+                "Resource": "*",
+                "Sid": "Karpenter"
+            },
+            {
+                "Action": "ec2:TerminateInstances",
+                "Condition": {
+                    "StringLike": {
+                        "ec2:ResourceTag/Name": "*karpenter*"
+                    }
+                },
+                "Effect": "Allow",
+                "Resource": "*",
+                "Sid": "ConditionalEC2Termination"
+            }
+        ],
+        "Version": "2012-10-17"
+    }"#,
+        )
+        .send()
+        .await
+        .context(
+            Resources::Clear,
+            "Unable to create Karpenter controller policy",
+        )?;
+    Ok(())
+}
+
+/// This is the object that will destroy ec2 instances.
+pub struct Ec2KarpenterDestroyer {}
+
+#[async_trait::async_trait]
+impl Destroy for Ec2KarpenterDestroyer {
```

Comment:
Does this clean up everything? Or does it leave some things behind?

Reply:
It cleans up all of the conflicting things. You can run this agent with the regular ec2 agent and everything works. In the `future work` section (which will become new issues) I've included improving cleanup.